File: app/YetiDoc/bs/src/module.ml
(*  Copyright 26-Jul-2018 ºDeme
    GNU General Public License - V3 <http://www.gnu.org/licenses/>
*)

open Main
open Ui


let fsort s1 s2 =
  let up = Js.String.toUpperCase in
  String.compare (up s1) (up s2)

let linkTds tp a =
  let a = a |> It.of_array |> It.sort fsort |> It.to_array in
  let len = Array.length a in
  let cols = 4 in
  let rows = (len - 1) / cols + 1 in
  It.reduce []
    (fun ls row ->
      q "tr" []
        (It.reduce []
          (fun ls col ->
            let ix = col * rows + row in
            if ix >= len then q "td" [][]::ls
            else q "td" [
              Html ("<a href='#" ^ tp ^ "." ^ a.(ix) ^ "'>" ^ a.(ix) ^ "</a>")
            ][]::ls
          )
          (It.range 0 cols) |>
          List.rev)
      :: ls
    )
    (It.range 0 rows) |>
    List.rev

let groupTds gr tp a =
  match a with
  | [||] -> []
  | _ ->
    q "tr" [] [
      q "td" [Att_i ("colspan", 4); Html ("<i>" ^ gr ^ "</i>")] []]::
    linkTds tp a

let index tree =
  let head tp =
    q "tr" [] [q "td" [Att_i ("colspan", 4)][q "hr" [][]]]::
    q "tr" [] [
      q "td" [
        Att_i ("colspan", 4);
        Html (if tp == "~"
             then "<b>Module signature</b>"
             else("<b>Type " ^ tp ^ "</b>"))
      ][]]::
    [] in
  let entries e = Module_data.(
    (head e.tp) @
    (groupTds "Variant" e.tp e.enums) @
    (groupTds "Parameters" e.tp e.ps) @
    (groupTds "Functions" e.tp e.ms)) in
  let tds = Array.fold_left (fun s e -> s @ (entries e)) [] tree in
  q "table" [Klass "main"] tds

let overview d =
  let (name, link) = Module_data.(d.hyperlink) in
  q "div" [][
    q "hr" [][];
    q "p" [][
      q "span" [Html "<b>File</b>"][];
      q "br" [][];
      q "a" [Att ("href", link); Html name][]];
    q "p" [Klass "frame"; Html "<b>Overview</b>"][]]


let bottom = q "div" []
  It.(map (fun _ -> q "p" [Html "&nbsp;"][]) (range 1 28) |> to_list)

let up_arrow =
  q "div" [Style "position: fixed;bottom: 0px;right: 20px"][
    link (fun _ -> location_assign "#") |> Domo.add [img "up"]]

let scroll_to_tag : unit -> unit =
  [%raw {|
    function () {
      let hash = location.hash;
      hash = hash.substring(1, hash.length);
      if (hash !== null)
        document.getElementById(hash).scrollIntoView();
    }
  |}]

let show' mpath d =
  let (title, html1, html2, tree) =
    Module_data.((d.title, d.html1, d.html2, d.tree)) in
  let w =
    q "div" [][
      q "div" [Html html1][];
      q "div" [][index tree];
      q "div" [][overview d];
      q "div" [Html html2][
        bottom;
        up_arrow;]]
  in (
    show mpath w;
    let _ = (It.get (qq "title"))
      |> Domo.set [Text title] in ();
    scroll_to_tag ()
  )

let show mpath fpath =
  let rq = Json.(Client.wrq [
    "page", wstring "module";
    "mpath", wstring mpath;
    "fpath", wstring fpath;
  ])
  in
  Client.send (client ()) rq (fun rp ->
      match Client.rrp rp "data" (Json.ropt Module_data.of_json
      (*Module_data.of_json*)) with
      | None ->
        let rq = Json.(Client.wrq [
          "page", wstring "index";  (* Page index is reused *)
          "rq", wstring "setMenu";
          "option", wstring "@";
        ])
        in
        Client.send (client ()) rq (fun _ -> location_assign "?@")
      | Some d ->
        let total_menu = mpath ^ "@" ^ fpath in
        let rq = Json.(Client.wrq [
          "page", wstring "index";  (* Page index is reused *)
          "rq", wstring "setMenu";
          "option", wstring total_menu;
        ])
        in
        Client.send (client ()) rq (fun _ -> show' mpath d)
    )

File: app/YetiDoc/js/src/Module.js
// Copyright 02-Feb-2020 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

import Client from "./dmjs/Client.js"; // eslint-disable-line
import Ui from "./dmjs/Ui.js";
import Domo from "./dmjs/Domo.js"; // eslint-disable-line
import Maybe from "./dmjs/Maybe.js";
import {Mod} from "./data/Mod.js";

const $ = e => Ui.$(e);

/**
    Index page.
**/
export default class Module {

  /**
      @param {!Client} client
      @param {string} lib
      @param {string} modPath
      @param {!Array<?>} d
  **/
  constructor (client, lib, modPath, d) {
    this._client = client;
    this._lib = lib;
    this._modPath = modPath;
    this._d = Mod.fromJs(d);

    this._wg = $("div");
    this.view();
  }

  /**
      @return {!Domo}
  **/
  get wg () {
    return this._wg;
  }


  // View ----------------------------------------------------------------------
  /**
      @private
      @return {void}
  **/
  view () {
    const d = this._d;

    $("@title").text(this._modPath);
    this._wg.removeAll()
      .add($("div").html(d.html1))
      .add(this.index(d.tree))
      .add(this.overview(d.link))
      .add($("div").html(d.html2))
      .add($("div").style("height:1200px"))
      .add(Ui.upTop("up"))
    ;
  }

  index (tree) {
    function head (tp) {
      return [
        $("tr").add($("td").att("colspan", 4).add($("hr"))),
        $("tr")
          .add($("td").att("colspan", 4)
            .html(tp === "~"
              ? "<b>Module signature</b>"
              : "<b>Type</b> " + tp
            ))
      ];
    }

    function rows (tp, elements) {
      elements.sort((e1, e2) => e1.toUpperCase() > e2.toUpperCase() ? 1 : 0);
      const len = elements.length;
      const cols = 4;
      const rows = ((len + cols - 1) / cols) | 0;
      const rs = [];
      for (let r = 0; r < rows; ++r) {
        const row = $("tr");
        for (let c = 0; c < cols; ++c) {
          const ix = c * rows + r;
          if (ix >= len) {
            row.add($("td"));
          } else {
            row.add($("td").html(
              "<a href=\"#" + tp + "." + elements[ix] + "\">" +
              elements[ix] + "</a>"
            ));
          }
        }
        rs.push(row);
      }
      return rs;
    }

    function groupTrs (group, tp, elements) {
      if (elements.length === 0) return [];
      const r = [
        $("tr").add($("td").att("colspan", 4).html("<i>" + group + "</i>"))
      ];
      for (const row of rows(tp, elements)) {
        r.push(row);
      }
      return r;
    }

    function index (e) {
      if (e.enums.length + e.ps.length + e.ms.length === 0) return [];

      return head(e.tp)
        .concat(groupTrs("Variants", e.tp, e.enums))
        .concat(groupTrs("Fields", e.tp, e.ps))
        .concat(groupTrs("Functions", e.tp, e.ms))
      ;
    }

    function tds () {
      const r = [];
      for (const e of tree) {
        for (const row of index(e)) {
          r.push(row);
        }
      }
      return r;
    }

    return $("table").klass("main")
      .adds(tds());
  }

  overview (link) {
    return $("div")
      .add($("hr"))
      .add($("p")
        .add($("span").html("<b>File</b>"))
        .add($("br"))
        .add($("a").att("href", link.link).html(link.name)))
      .add($("p").klass("frame").html("<b>Overview</b>"))
    ;
  }

  // Control -------------------------------------------------------------------

  /**
      @param {!Client} client
      @param {string} lib
      @param {string} libPath
      @param {string} modPath
      @return {!Promise<!Maybe<!Module>>}
  **/
  static async mk (client, lib, libPath, modPath) {
    const rp = await client.send({
      "source": "Module",
      "lib": lib,
      "libPath": libPath,
      "modPath": modPath
    });
    return Maybe.fromJs(rp["data"]).fmap(d =>
      new Module(client, lib, modPath, d)
    );
  }

}
